packages/deploy/utils/lazyMinting.ts

```diff
@@ -0,0 +1,357 @@
+import {
+  AbiCoder,
+  BytesLike,
+  Contract,
+  Numeric,
+  Signer,
+  ZeroAddress,
+} from 'ethers';
+import {ethers} from 'hardhat';
+import {Network} from 'hardhat/types';
+
+export type LazyMintData = {
+  caller: string;
+  tier: bigint;
+  amount: bigint;
+  unitPrice: bigint;
+  paymentToken: string;
+  metadataHash: string;
+  maxSupply: bigint;
+  creator: string;
+};
+
+export type LazyMintBatchData = {
+  caller: string;
+  tiers: bigint[];
+  amounts: bigint[];
+  unitPrices: bigint[];
+  paymentTokens: string[];
+  metadataHashes: string[];
+  maxSupplies: bigint[];
+  creators: string[];
+};
+
+export type Order = {
+  maker: string;
+  makeAsset: Asset;
+  taker: string;
+  takeAsset: Asset;
+  salt: Numeric;
+  start: Numeric;
+  end: Numeric;
+};
+
+export enum AssetClassType {
+  INVALID_ASSET_CLASS = '0x0',
+  ERC20_ASSET_CLASS = '0x1',
+  ERC721_ASSET_CLASS = '0x2',
+  ERC1155_ASSET_CLASS = '0x3',
+}
+
+export type Asset = {
+  assetType: AssetType;
+  value: Numeric;
+};
+
+export type AssetType = {
+  assetClass: AssetClassType;
+  data: BytesLike;
+};
+
+export const AssetERC1155 = async (
+  tokenContract: Contract,
+  tokenId: Numeric,
+  value: Numeric
+): Promise<Asset> => ({
+  assetType: {
+    assetClass: AssetClassType.ERC1155_ASSET_CLASS,
+    data: AbiCoder.defaultAbiCoder().encode(
+      ['address', 'uint256'],
+      [await tokenContract.getAddress(), tokenId]
+    ),
+  },
+  value,
+});
+
+export const AssetERC20 = async (
+  tokenContract: Contract,
+  value: Numeric
+): Promise<Asset> => ({
+  assetType: {
+    assetClass: AssetClassType.ERC20_ASSET_CLASS,
+    data: AbiCoder.defaultAbiCoder().encode(
+      ['address'],
+      [await tokenContract.getAddress()]
+    ),
+  },
+  value,
+});
+
+export async function signOrder(
+  order: Order,
+  account: Signer,
+  verifyingContract: Contract
+) {
+  return account.signTypedData(
+    {
+      name: 'The Sandbox Marketplace',
+      version: '1.0.0',
```

Comment:
Should we increment the version to match the package at all? @mvanmeerbeck

Reply:
@atkinsonholly seems this value is hardcoded in the `OrderValidator.sol` and was not updated for marketplace 1.0.2 so I kept it as is.

Reply:
maybe, doesn't matter right now, we're using 1.0.0 anyway